### PR TITLE
SLING-12115 - Repoinit should leave importBehaviour for ACL creation to JCR

### PR DESCRIPTION
**Pull request by jsedding**

### CODE REVIEW

File: src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java [hunk 1]
```diff
@@ -156,8 +156,11 @@ private static void setAcl(Session session, List<String> principals, String jcrP
             if (principal == null) {
                 // backwards compatibility: fallback to original code treating principal name as authorizable ID (see SLING-8604)
                 final Authorizable authorizable = UserUtil.getAuthorizable(session, name);
-                checkState(authorizable != null, "Authorizable not found: {0}", name);
-                principal = authorizable.getPrincipal();
+                if (authorizable != null) {
+                    principal = authorizable.getPrincipal();
+                } else {
+                    principal = () -> name;
```
Comment (anchela) [inline src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:162]: since the previous behavior was to throw an exception, i would recommend to at least log a warning here.

**Review by anchela — CHANGES_REQUESTED** [review]
hi julian, from an oak point of view this change is sensible (and actually aligns with the behavior of content packages). just a minor wish (add some log output) prevents me from approving right away.

if i look at it from a pure sling point of view and think about potential implications one could argue that this may lead to unintended ac-setup for non-existing principals being writting that up to now was prevented. maybe that's not a valid concern, but i want to mention it for the sake of completeness.

it would recommend though to get a second opinion from someone looking at this without oak-blindness. e.g. @cziegeler , @rombert  or @kwin

**Comment by jsedding** [comment]
Thinking about this change some more, I am inclined to close this PR. If we really want to support creating ACLs for possible non-existing principals, we should add that to the repoinit language and make it explicit.

**Comment by anchela** [comment]
@jsedding , i am not entirely sure that this modification justifies adding new repo-init language. but if we take that route, please make also sure to adjust other bundles like the cp-converter. it's doable but if forgotten will result in breaking AEMaaCS.

**Comment by kwin** [comment]
I think this doesn't require a new instruction but just an extension for  principal centered ACLs with options (introduced in https://issues.apache.org/jira/browse/SLING-6423).

**Comment by jsedding** [comment]
I like @kwin's suggestion to use options. Especially if that does not require a parser change. Would that work for you @anchela?

@anchela yes, I am aware of the (or at least some) complications of updating the repoinit parser. I'll exercise caution.

**Comment by anchela** [comment]
@jsedding , the options were initially introduced into the parser to add the same option-nightmare that we are having in filevault. fortunately, this was never implemented in jcr-repoinit and i hoped to make it clear in JIRA why I would oppose to that.
if we can repurpose the options for something sensible, i don't mind.... but maybe you also want to reach out @bdelacretaz . he might want to have a say here as the original author of repoinit.

**Comment by jsedding** [comment]
I have implemented this using the following, option based, syntax:

```
set ACL for nonExistingGroup (ACLOptions=ignoreMissingPrincipal)
  allow jcr:read on /
end
```

With this implementation, the current behaviour remains unchanged. Only when this new option is present, will the check for existence of the principal in repoinit be skipped. In that case, the behaviour of the repository determines, what happens.

Does this look acceptable? And could this approach be recommended to be employed for cases where an external principal is not yet sync'ed into the repository?

cc @bdelacretaz @anchela @kwin

**Comment by jsedding** [comment]
If there are no objections, I will merge this PR tomorrow.

File: src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java (hunk 1 above)
Comment (jsedding) [reply src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java]: I think this is resolved with my latest changes.

File: src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java [hunk 2]
```diff
@@ -152,7 +154,9 @@ private static void setAcl(Session session, List<String> principals, String jcrP
 
         boolean changed = false;
         for (String name : principals) {
-            Principal principal = AccessControlUtils.getPrincipal(session, name);
+            Principal principal = options.contains("ignoreMissingPrincipal")
+                    ? () -> name
```
Comment (anchela) [inline src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:158]: i would slightly alter the logic here, as the fallback below is significant..... if you just use the specified name now, you might break that contract below.

so, i would suggest:
- always try to retrieve the principal first
- if it does not exist -> try backwards compatible fallback to original code
- if it does not exist -> create the anonymous principal in case option is set as you do here, otherwise use old behavior.

i believe this way we run the least risk of introducing subtle regressions that are super hard to troubleshoot.

File: src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java [hunk 3]
```diff
@@ -420,7 +419,7 @@ private static boolean isValidPath(@NotNull Session session, @Nullable String jc
     /**
      * 
      * @param acMgr the access control manager
-     * @param principal the principal
+     * @param path the path
```
Comment (anchela) [inline src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:422]: is that change related?

File: src/main/java/org/apache/sling/jcr/repoinit/impl/AclVisitor.java
Comment (anchela) [inline src/main/java/org/apache/sling/jcr/repoinit/impl/AclVisitor.java:1]: sorry too many changes to review with reasonable effort :(
can you avoid any renaming of methods that are unrelated to the change required here?
i am all in favor of cleaning up the code base, but please use a separate ticket for that.

File: src/main/java/org/apache/sling/jcr/repoinit/impl/DoNothingVisitor.java [hunk 4]
```diff
@@ -66,7 +66,7 @@ protected DoNothingVisitor(Session s) {
         session = s;
     }
 
-    protected void report(Exception e, String message) {
+    protected static void report(Exception e, String message) {
```
Comment (anchela) [inline src/main/java/org/apache/sling/jcr/repoinit/impl/DoNothingVisitor.java:69]: unrelated to the task at hand. please move that to a general cleanup.improvmeent PR.

File: src/test/java/org/apache/sling/jcr/repoinit/GeneralAclTest.java [hunk 5]
```diff
@@ -270,9 +270,9 @@ private void verifyRegisterNamespace(String username, String prefix, String uri,
         try {
             userSession = U.loginService(username);
             userSession.getWorkspace().getNamespaceRegistry().registerNamespace(prefix, uri);
-            assertTrue("Register namespace succeeded " + prefix + uri, successExpected);
+            assertTrue("Register namespace succeeded " + prefix + " " + uri, successExpected);
         } catch (RepositoryException e) {
-            assertTrue("Error registering namespace " + prefix + uri + " " + e.getMessage(), !successExpected);
+            assertFalse("Error registering namespace " + prefix + " " + uri + " " + e.getMessage(), successExpected);
```
Comment (anchela) [inline src/test/java/org/apache/sling/jcr/repoinit/GeneralAclTest.java:275]: ???? what does the changes for SLING-12115 lead to a changed assertion in a namespace related test? that doesn't look good at all.

**Review by anchela — CHANGES_REQUESTED** [review]
@jsedding , i reviewed it again and i would like you to make a few changes.

File: src/main/java/org/apache/sling/jcr/repoinit/impl/AclVisitor.java
Comment (jsedding) [reply src/main/java/org/apache/sling/jcr/repoinit/impl/AclVisitor.java:1]: I did this refactoring so that `SetAclPrincipals` and `SetAclPaths` share the same code-path and to remove code-duplication. For symmetry, and because just removing the "set ACL" part from the method `setAcl` seemed strange, I did the same for `RemoveAcePrincipals` and `RemoveAcePaths`. The final reason, why I didn't want a `setAcl` method in this class is, that it clashes with `AclUtil#setAcl`, which I found confusing.

The method `setRepositoryAcl` was a copy & paste of `setAcl`, omitting only the `paths` argument, which was set to `null` in the removed `AclUtil#setRepositoryAcl`. Removing this in favour of explicitly injecting the `:repository` path, seemed much clearer to me.

Finally, the `require` method was actually a no-op, because `AclLine#getProperty()` [never returns `null`](https://github.com/apache/sling-org-apache-sling-repoinit-parser/blob/ea340191058fefe84ba9134f8025f71a69295a14/src/main/java/org/apache/sling/repoinit/parser/operations/AclLine.java#L64-L67).

If you insist, I can try to revert these changes in favour of a minimal change-set approach.

File: src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java (hunk 3 above)
Comment (jsedding) [reply src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:422]: Nope, just javadoc cleanup. I'll revert it.

File: src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java (hunk 2 above)
Comment (jsedding) [reply src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:158]: Sounds sensible. Thanks for catching this change in logic!

File: src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java [hunk 6]
```diff
@@ -152,7 +154,9 @@ private static void setAcl(Session session, List<String> principals, String jcrP
 
         boolean changed = false;
         for (String name : principals) {
-            Principal principal = AccessControlUtils.getPrincipal(session, name);
+            Principal principal = options.contains("ignoreMissingPrincipal")
```
[the remark this answers is not in the document]
Comment (jsedding) [reply src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:157]: +1 to both points.

Regarding documentation, it's probably best to add a test-case example in the repoinit parser module. AFAIK the test files are included in the documentation. I suppose that's the downside of having `ACLOptions` accept all strings.

File: src/main/java/org/apache/sling/jcr/repoinit/impl/DoNothingVisitor.java (hunk 4 above)
Comment (jsedding) [reply src/main/java/org/apache/sling/jcr/repoinit/impl/DoNothingVisitor.java:69]: +1

File: src/test/java/org/apache/sling/jcr/repoinit/GeneralAclTest.java (hunk 5 above)
Comment (jsedding) [reply src/test/java/org/apache/sling/jcr/repoinit/GeneralAclTest.java:275]: True, that's cleanup.

**Comment by jsedding** [comment]
@anchela thank you for the review! I have tried to address the issues you pointed out, and I have tried to simplify the diff of the AclVisitor class, without reverting back to the previous version. I'll let you be the judge of whether I succeeded in simplifying the diff 😅

File: src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java (hunk 1 above)
Comment (anchela) [reply src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java]: yes, i agree

File: src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (anchela) [reply src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java]: lgtm now

File: src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java (hunk 3 above)
Comment (anchela) [reply src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java]: don;t get me wrong..... i am all for cleaning up code and actively maintaining it... but i would love this to happen with a separate PR. such that we can really track what has changed and why.

File: src/main/java/org/apache/sling/jcr/repoinit/impl/AclVisitor.java
Comment (anchela) [reply src/main/java/org/apache/sling/jcr/repoinit/impl/AclVisitor.java:1]: @jsedding , those are all valid points... it's just that i don't have the bandwidths to review it. too many reviews piling up. i don't insist that you revert it and i see the benefit of improving the repoinit code.... but i would love someone else to take a look and verify that the result is equivalent. maybe @kwin , @cziegeler  or @rombert can jump in here?

File: src/main/java/org/apache/sling/jcr/repoinit/impl/DoNothingVisitor.java (hunk 4 above)
Comment (anchela) [reply src/main/java/org/apache/sling/jcr/repoinit/impl/DoNothingVisitor.java]: i see

File: src/test/java/org/apache/sling/jcr/repoinit/GeneralAclTest.java (hunk 5 above)
Comment (anchela) [reply src/test/java/org/apache/sling/jcr/repoinit/GeneralAclTest.java]: but unrelated to the task at hand, right?
i really would like changes to a particular bug or improvement to be limited to the task at hand.... for me just looking at the diff it's just harder to figure out why something was change.

**Review by anchela — COMMENTED** [review]
i reviewed the changes with exception of the code refactoring. someone else should take a look as i don't have enough bandwidth for that. so leaving the approval for those.
i can't reset my 'request changes' though without approving it.... so feel free to ignore it.

**Comment by jsedding** [comment]
Thank you for your reviews, @anchela! Addressing the points you raised forced me to look at all the details again, and I believe the result is an improvement over the previous state.

File: src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java [hunk 7]
```diff
@@ -174,9 +173,26 @@ private static void setAcl(Session session, List<String> principals, String jcrP
...
-    public static void setRepositoryAcl(Session session, List<String> principals, List<String> privileges, boolean isAllow, List<RestrictionClause> restrictionClauses)
-           throws RepositoryException {
-        setAcl(session, principals, (String)null, privileges, isAllow, restrictionClauses);
+    @NotNull
+    private static Principal getPrincipal(Session session, String name, boolean ignoreMissingPrincipal) throws RepositoryException {
+        Principal principal = AccessControlUtils.getPrincipal(session, name);
+        if (principal == null) {
+            // backwards compatibility: fallback to original code treating principal name as authorizable ID (see SLING-8604)
+            final Authorizable authorizable = UserUtil.getAuthorizable(session, name);
+            if (!ignoreMissingPrincipal) {
```
Comment (rombert) [inline src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:182]: Nit: I would remove the double negation and write the statement as

if ( ignoreMissingPrincipal ) {
...
} else {
...
}

**Review by rombert — APPROVED** [review]
Looks good to me overall

File: src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java (hunk 7 above)
Comment (jsedding) [reply src/main/java/org/apache/sling/jcr/repoinit/impl/AclUtil.java:182]: I prefer it the way it is. Granted, the condition _is_ probably harder to read and understand with what is effectively a double negation. For me, having what I consider to be the default code-path first, makes reading the code easier again. But I can totally see that there's a good argument to be made for both ways.

**Comment by jsedding** [comment]
@anchela in order to clear your "changes request" status from the PR, I clicked the "re-request review" icon. Should have done that yesterday, before you reviewed again.
